Query for implementation of `StreamFieldInterface` breaks preview

I just discovered that GraphQL queries involving the fragments for implementations of the `StreamFieldInterface` (like `TextBlock` or `ImageChooserBlock`) break the functionality of the preview feature. 

A query like this
```graphql
query ($slug: String) {
  wagtail {
    page(slug: $slug) {
      ... on BlogPage {
        id
        title
        freeformbody {  # this is the StreamField
          id
        }
      }
    }
  }
}
```
works. But as soon as I add a implementation fragment like so 
```graphql
query ($slug: String) {
  wagtail {
    page(slug: $slug) {
      ... on BlogPage {
        id
        title
        freeformbody {
          id
          ... on TextBlock {
            rawValue
          }
        }
      }
    }
  }
}
```
then the preview page remains empty. Looking at the developer console I can find the following error messages: 
> You are using the simple (heuristic) fragment matcher, but your queries contain union or interface types. Apollo Client will not be able to accurately map fragments. To make this error go away, use the `IntrospectionFragmentMatcher` as described in the docs: https://www.apollographql.com/docs/react/advanced/fragments.html#fragment-matcher

> WARNING: heuristic fragment matching going on!

The queries are fine in the GraphiQL interface and the Gatsby development and build pages work too. 

The link in the error message itself only leads to a 404 page. I guess the Apollo documentation has been restructured since. I tried to search for keywords from the error message in the documentation and found the following two resources.
* [Configuring the cache - Client (React) - Apollo GraphQL Docs](https://www.apollographql.com/docs/react/v2.6/caching/cache-configuration/)
* [Using fragments - Client (React) - Apollo GraphQL Docs](https://www.apollographql.com/docs/react/v2.6/data/fragments/#fragments-on-unions-and-interfaces)

It appears that something along the lines of this
```js
import { IntrospectionFragmentMatcher } from 'apollo-cache-inmemory';
import introspectionQueryResultData from './fragmentTypes.json';

... 

const fragmentMatcher = new IntrospectionFragmentMatcher({
  introspectionQueryResultData
});

const cache = new InMemoryCache({ fragmentMatcher });

const client = new ApolloClient({
  cache,
  link: new HttpLink(),
});
```
would need to be added [in the preview.boilerplate.js](https://github.com/GrappleGQL/gatsby-source-wagtail/blob/4f85ecff145a4cf68583abdc4f82e8691b496a81/src/preview.boilerplate.js#L298). 

I will give this a shot, but I have to say that I am not experienced with Gatsby plugin nor NPM package development. 

In case it is really that simple and solves the issue, I am happy to create a PR, but I assume that this will require some more experienced developers input.

As I was afraid of, I am not even getting an example site running with a local copy of the plugin. 

I have tried all three setups mentioned in [Creating a Local Plugin | GatsbyJS](https://www.gatsbyjs.org/docs/creating-a-local-plugin/). 
Neither a local `/plugins` folder nor `resolve.require` in the `gatsby-config.js` nor `npm link ...` work for me. 

The first two options result in this error message:
```
...
  Error: Cannot find module 'gatsby-source-wagtail/pages.js'
...
```
The file does exist though. Cleaning the plugin directory and the page directory and running `npm run build` or `npm run watch` in the plugin repo also have no effect on this. 

The `npm link ...` option results in a different error:
```
...
 ERROR

Missing onError handler for invocation 'building-schema', error was 'Error: Cannot create as TypeComposer the following value: Query.'. Stacktrace was 'Error: Cannot create as TypeComposer the following value: Query.
    at SchemaComposer.createTempTC (/Users/tibor/1-Projects/gatsby-source-wagtail-preview-issue/example-frontend/node_modules/graphql-compose/lib/SchemaComposer.js:365:11)
    at /Users/tibor/1-Projects/gatsby-source-wagtail-preview-issue/example-frontend/node_modules/gatsby/dist/schema/schema.js:708:36
    at Array.forEach (<anonymous>)
    at addThirdPartySchemas (/Users/tibor/1-Projects/gatsby-source-wagtail-preview-issue/example-frontend/node_modules/gatsby/dist/schema/schema.js:706:21)
    at updateSchemaComposer (/Users/tibor/1-Projects/gatsby-source-wagtail-preview-issue/example-frontend/node_modules/gatsby/dist/schema/schema.js:210:9)
    at async buildSchema (/Users/tibor/1-Projects/gatsby-source-wagtail-preview-issue/example-frontend/node_modules/gatsby/dist/schema/schema.js:91:3)
    at async build (/Users/tibor/1-Projects/gatsby-source-wagtail-preview-issue/example-frontend/node_modules/gatsby/dist/schema/index.js:141:18)
    at async buildSchema (/Users/tibor/1-Projects/gatsby-source-wagtail-preview-issue/example-frontend/node_modules/gatsby/dist/services/build-schema.js:20:3)'

...
``` 

Not sure how to proceed here. 

@NathHorrigan @zerolab How do you set up the development for this package? 

Also, can you even reproduce the issue?

--- a/package.json
+++ b/package.json
@@ -22,6 +22,7 @@
     "apollo-utilities": "^1.3.4",
     "fs-extra": "^8.1.0",
     "gatsby": "^2.13.77",
+    "gatsby-source-filesystem": "^2.3.20",
     "gatsby-source-graphql": "^2.1.8",
     "graphql": "^14.5.0",
     "graphql-2-json-schema": "^0.2.0",

--- a/src/preview.boilerplate.js
+++ b/src/preview.boilerplate.js
@@ -7,14 +7,17 @@ import traverse from 'traverse'
 import { ApolloClient } from 'apollo-client'
 import { gql } from 'apollo-boost'
 import { split } from 'apollo-link'
-import { InMemoryCache, IntrospectionFragmentMatcher } from 'apollo-cache-inmemory'
+import {
+    InMemoryCache,
+    IntrospectionFragmentMatcher,
+} from 'apollo-cache-inmemory'
 import { WebSocketLink } from 'apollo-link-ws'
 import { getMainDefinition } from 'apollo-utilities'
 import { createHttpLink } from 'apollo-link-http'
 import {
     introspectSchema,
     makeRemoteExecutableSchema,
-    mergeSchemas
+    mergeSchemas,
 } from 'graphql-tools'
 
 import { print } from 'graphql/language/printer'
@@ -28,7 +31,7 @@ const PreviewProvider = async (query, fragments = '', onNext) => {
         fieldName,
         url,
         websocketUrl,
-        headers
+        headers,
     } = window.___wagtail.default
     const isolatedQuery = getIsolatedQuery(query, fieldName, typeName)
     const { content_type, token } = decodePreviewUrl()
@@ -44,8 +47,8 @@ const PreviewProvider = async (query, fragments = '', onNext) => {
     let link = createHttpLink({
         uri: url,
         fetchOptions: {
-            headers: { Authorization: token ? `Basic ${getToken()}` : '' }
-        }
+            headers: { Authorization: token ? `Basic ${getToken()}` : '' },
+        },
     })
 
     // If provided create a subscription endpoint
@@ -56,9 +59,9 @@ const PreviewProvider = async (query, fragments = '', onNext) => {
             options: {
                 reconnect: true,
                 connectionParams: {
-                    authToken: getToken()
-                }
-            }
+                    authToken: getToken(),
+                },
+            },
         })
 
         // Alias original link and create one that merges the two
@@ -184,7 +187,7 @@ const PreviewProvider = async (query, fragments = '', onNext) => {
         return {
             imageHeight: Number(imageHeight),
             imageWidth: Number(imageWidth),
-            aspectRatio
+            aspectRatio,
         }
     }
 
@@ -195,7 +198,7 @@ const PreviewProvider = async (query, fragments = '', onNext) => {
                 const {
                     imageWidth,
                     imageHeight,
-                    aspectRatio
+                    aspectRatio,
                 } = computeSharpSize(source, info)
                 return {
                     __typename: 'ImageSharpFixed',
@@ -209,7 +212,7 @@ const PreviewProvider = async (query, fragments = '', onNext) => {
                     srcSet: '',
                     srcWebp: '',
                     srcSetWebp: '',
-                    originalName: ''
+                    originalName: '',
                 }
             },
             fluid: (root, args, context, info) => {
@@ -217,7 +220,7 @@ const PreviewProvider = async (query, fragments = '', onNext) => {
                 const {
                     imageWidth,
                     imageHeight,
-                    aspectRatio
+                    aspectRatio,
                 } = computeSharpSize(source, info)
                 return {
                     __typename: 'ImageSharpFluid',
@@ -233,9 +236,9 @@ const PreviewProvider = async (query, fragments = '', onNext) => {
                     originalImg: source.src,
                     originalName: '',
                     presentationWidth: imageWidth,
-                    presentationHeight: imageHeight
+                    presentationHeight: imageHeight,
                 }
-            }
+            },
         },
         CustomImage: {
             imageFile: (source, args, context, info) => {
@@ -277,34 +280,34 @@ const PreviewProvider = async (query, fragments = '', onNext) => {
                         __typename: 'ImageSharp',
                         fluid: {
                             __typename: 'ImageSharpFluid',
-                            parent: source
+                            parent: source,
                         },
                         fixed: {
                             __typename: 'ImageSharpFixed',
-                            parent: source
+                            parent: source,
                         },
                         original: {
                             __typename: 'ImageSharpOriginal',
                             height: source.height,
                             width: source.width,
-                            src: source.src
-                        }
-                    }
+                            src: source.src,
+                        },
+                    },
                 }
-            }
-        }
+            },
+        },
     }
 
     // Create Apollo client
     const fragmentMatcher = new IntrospectionFragmentMatcher({
-        introspectionQueryResultData
+        introspectionQueryResultData,
     })
     const cache = new InMemoryCache({ fragmentMatcher })
     const client = new ApolloClient({
         cache,
         link,
         typeDefs,
-        resolvers: schemaExtensionResolvers
+        resolvers: schemaExtensionResolvers,
     })
 
     if (content_type && token) {
@@ -318,17 +321,17 @@ const PreviewProvider = async (query, fragments = '', onNext) => {
         // Get first version of preview to render the template
         client
             .query({ query: gql([query]) })
-            .then(result => onNext(result.data || {}))
+            .then((result) => onNext(result.data || {}))
         // Subscribe to any changes...
         client
             .subscribe({
                 query: gql([subscriptionQuery]),
-                variables: {}
+                variables: {},
             })
             .subscribe(
-                response => onNext(response),
-                error => console.log(error),
-                complete => console.log(complete)
+                (response) => onNext(response),
+                (error) => console.log(error),
+                (complete) => console.log(complete)
             )
     }
 }
@@ -339,11 +342,11 @@ export const withPreview = (WrappedComponent, pageQuery, fragments = '') => {
         constructor(props) {
             super(props)
             this.state = {
-                wagtail: cloneDeep(props.data ? props.data.wagtail : {})
+                wagtail: cloneDeep(props.data ? props.data.wagtail : {}),
             }
-            PreviewProvider(pageQuery, fragments, data => {
+            PreviewProvider(pageQuery, fragments, (data) => {
                 this.setState({
-                    wagtail: merge({}, this.state.wagtail, data)
+                    wagtail: merge({}, this.state.wagtail, data),
                 })
             })
         }
@@ -367,26 +370,26 @@ const generatePreviewQuery = (query, contentType, token, fragments) => {
             kind: 'Argument',
             name: {
                 kind: 'Name',
-                value: 'contentType'
+                value: 'contentType',
             },
             value: {
                 block: false,
                 kind: 'StringValue',
-                value: contentType
-            }
+                value: contentType,
+            },
         },
         {
             kind: 'Argument',
             name: {
                 kind: 'Name',
-                value: 'token'
+                value: 'token',
             },
             value: {
                 block: false,
                 kind: 'StringValue',
-                value: token
-            }
-        }
+                value: token,
+            },
+        },
     ]
 
     // Rename query for debugging reasons
@@ -395,14 +398,14 @@ const generatePreviewQuery = (query, contentType, token, fragments) => {
     queryDef.variableDefinitions = []
 
     // Add field to AST
-    const createSelection = name => ({
+    const createSelection = (name) => ({
         kind: 'Field',
         name: {
             kind: 'Name',
-            value: name
+            value: name,
         },
         arguments: [],
-        directives: []
+        directives: [],
     })
 
     // Alter the query so that we can execute it properly
@@ -412,7 +415,7 @@ const generatePreviewQuery = (query, contentType, token, fragments) => {
         if (
             node?.kind == 'Field' &&
             node?.selectionSet?.selections?.find(
-                selection =>
+                (selection) =>
                     selection?.name?.value == 'imageFile' &&
                     (imageFileNode = selection)
             )
@@ -429,21 +432,21 @@ const generatePreviewQuery = (query, contentType, token, fragments) => {
                 kind: 'Directive',
                 name: {
                     kind: 'Name',
-                    value: 'client'
-                }
+                    value: 'client',
+                },
             })
 
             // Replace inline any fragments
             const fragmentTypes = require('gatsby-transformer-sharp/src/fragments.js')
-            traverse(imageFileNode).map(node => {
+            traverse(imageFileNode).map((node) => {
                 if (
                     node?.name?.value == 'fixed' ||
                     node?.name?.value == 'fluid' ||
                     node?.name?.value == 'original'
                 ) {
                     node.selectionSet.selections = node.selectionSet.selections
-                        .map(selection => {
-                            Object.keys(fragmentTypes).map(fragmentName => {
+                        .map((selection) => {
+                            Object.keys(fragmentTypes).map((fragmentName) => {
                                 if (selection?.name?.value == fragmentName) {
                                     const mod = fragmentTypes[fragmentName]
                                     const selections = gql([mod.source])
@@ -454,7 +457,7 @@ const generatePreviewQuery = (query, contentType, token, fragments) => {
                             })
                             return selection
                         })
-                        .filter(selection => !!selection)
+                        .filter((selection) => !!selection)
                 }
             })
 
@@ -468,7 +471,7 @@ const generatePreviewQuery = (query, contentType, token, fragments) => {
     } else {
         queryDef.name = {
             kind: 'Name',
-            value: 'PreviewQuery'
+            value: 'PreviewQuery',
         }
     }
 
@@ -478,18 +481,18 @@ const generatePreviewQuery = (query, contentType, token, fragments) => {
     non-page selections so we override the whole array with just the pages.
   */
     const pageSelections = queryDef.selectionSet.selections.filter(
-        selection => {
+        (selection) => {
             return selection.name.value.toLowerCase() === 'page'
         }
     )
-    pageSelections.map(selection => (selection.arguments = previewArgs))
+    pageSelections.map((selection) => (selection.arguments = previewArgs))
 
     // Change query to subcription type
     const subscriptionQuery = cloneDeep(queryDef)
     subscriptionQuery.operation = 'subscription'
     subscriptionQuery.selectionSet.selections = pageSelections
 
-    const updateFragments = fragments => {
+    const updateFragments = (fragments) => {
         return fragments
             .replace('on ImageSharpFixed', 'on ImageSharpFixed @client')
             .replace('on ImageSharpFluid', 'on ImageSharpFluid @client')
@@ -498,7 +501,7 @@ const generatePreviewQuery = (query, contentType, token, fragments) => {
 
     return {
         query: `${updateFragments(fragments)} ${print(query)}`,
-        subscriptionQuery: `${fragments} ${print(subscriptionQuery)}`
+        subscriptionQuery: `${fragments} ${print(subscriptionQuery)}`,
     }
 }
 

--- a/src/preview.boilerplate.js
+++ b/src/preview.boilerplate.js
@@ -7,7 +7,7 @@ import traverse from 'traverse'
 import { ApolloClient } from 'apollo-client'
 import { gql } from 'apollo-boost'
 import { split } from 'apollo-link'
-import { InMemoryCache } from 'apollo-cache-inmemory'
+import { InMemoryCache, IntrospectionFragmentMatcher } from 'apollo-cache-inmemory'
 import { WebSocketLink } from 'apollo-link-ws'
 import { getMainDefinition } from 'apollo-utilities'
 import { createHttpLink } from 'apollo-link-http'
@@ -296,8 +296,12 @@ const PreviewProvider = async (query, fragments = '', onNext) => {
     }
 
     // Create Apollo client
+    const fragmentMatcher = new IntrospectionFragmentMatcher({
+        introspectionQueryResultData
+    })
+    const cache = new InMemoryCache({ fragmentMatcher })
     const client = new ApolloClient({
-        cache: new InMemoryCache(),
+        cache,
         link,
         typeDefs,
         resolvers: schemaExtensionResolvers

